src/main/java/io/strimzi/kproxy/interceptor/ProduceRecordMutatingInterceptor.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.strimzi.kproxy.interceptor;
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.message.ProduceRequestData.PartitionProduceData;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.record.CompressionType;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
+import org.apache.kafka.common.record.MutableRecordBatch;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.TimestampType;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandler;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.strimzi.kproxy.codec.DecodedRequestFrame;
+
+/**
+ * An interceptor for modifying the key/value/header/topic of {@link ApiKeys#PRODUCE} requests.
+ */
+public class ProduceRecordMutatingInterceptor implements Interceptor {
+
+    @FunctionalInterface
+    public interface ByteBufferMutation {
+        ByteBuffer mutate(ByteBuffer original);
```

Comment:
If it is really mutating wouldn't a `void` return type by more obvious?

Reply:
Yeah, it's not mutating in place. Happy about any alternative proposals which don't create that expectation. `modify()`, `convert()`, `alter()`?